app/upgrade_cudos.go

```diff
@@ -585,76 +592,121 @@ func parseGenesisAccounts(jsonData map[string]interface{}, contractAccountMap *O
 	return accountMap, nil
 }
 
-func parseGenesisDelegations(validators *OrderedMap[string, *ValidatorInfo], contracts *OrderedMap[string, *ContractInfo], cudosCfg *CudosMergeConfig) (*OrderedMap[string, *OrderedMap[string, sdk.Int]], error) {
+func parseGenesisDelegations(validators *OrderedMap[string, *ValidatorInfo], contracts *OrderedMap[string, *ContractInfo], cudosCfg *CudosMergeConfig) (*OrderedMap[string, *OrderedMap[string, sdk.Int]], *OrderedMap[string, *OrderedMap[string, sdk.Int]], error) {
 	// Handle delegations
 	delegatedBalanceMap := NewOrderedMap[string, *OrderedMap[string, sdk.Int]]()
+	unbondingDelegatedBalanceMap := NewOrderedMap[string, *OrderedMap[string, sdk.Int]]()
+
 	for i := range validators.Iterate() {
 		validatorOperatorAddress, validator := i.Key, i.Value
 
-		for j := range validator.delegations.Iterate() {
+		for j := range validator.Delegations.Iterate() {
 			delegatorAddress, delegation := j.Key, j.Value
 
 			resolvedDelegatorAddress, err := resolveIfContractAddressWithFallback(delegatorAddress, contracts, cudosCfg)
 			if err != nil {
-				return nil, err
+				return nil, nil, err
 			}
 
 			currentValidatorInfo := validators.MustGet(validatorOperatorAddress)
-			delegatorTokens := currentValidatorInfo.TokensFromShares(delegation.shares).TruncateInt()
+			delegatorTokens := currentValidatorInfo.TokensFromShares(delegation.Shares).TruncateInt()
 
 			if delegatorTokens.IsZero() {
 				// This happens when number of shares is less than 1
 				continue
 			}
 
 			// Subtract balance from bonded or not-bonded pool
-			if currentValidatorInfo.status == BondedStatus {
+			if currentValidatorInfo.Status == BondedStatus {
 
 				// Store delegation to delegated map
 				resolvedDelegatorMap, _ := delegatedBalanceMap.GetOrSetDefault(resolvedDelegatorAddress, NewOrderedMap[string, sdk.Int]())
 				resolvedDelegator, _ := resolvedDelegatorMap.GetOrSetDefault(validatorOperatorAddress, sdk.NewInt(0))
 				resolvedDelegatorMap.Set(validatorOperatorAddress, resolvedDelegator.Add(delegatorTokens))
 				delegatedBalanceMap.Set(resolvedDelegatorAddress, resolvedDelegatorMap)
+			} else {
+				// Store delegation to delegated map
+				resolvedDelegatorMap, _ := unbondingDelegatedBalanceMap.GetOrSetDefault(resolvedDelegatorAddress, NewOrderedMap[string, sdk.Int]())
+				resolvedDelegator, _ := resolvedDelegatorMap.GetOrSetDefault(validatorOperatorAddress, sdk.NewInt(0))
+				resolvedDelegatorMap.Set(validatorOperatorAddress, resolvedDelegator.Add(delegatorTokens))
+				unbondingDelegatedBalanceMap.Set(resolvedDelegatorAddress, resolvedDelegatorMap)
+			}
```

Comment:
Shouldn't we have another if-else branch + map for handling unbonded (`BOND_STATUS_UNBONDED`) delegations.
👉 I understand that we consider `BOND_STATUS_UNBONDED` delegations as liquid balance, just wondering if it would have sense to have them stored in dedicated map, so we could potentially save them as list (per delegator account) in to the manifest file.
👉 Potentiall ythe same could apply for more esoteric `BOND_STATUS_UNSPECIFIED` delegation state (though I'm not entirely sure, whether this bond state actualy applies exclusivelly for validator state, though something tells me that if validator is in the `BOND_STATUS_UNSPECIFIED` state, all delegations on that validator will be in that state as well.

Reply:
Each delegator has list of `Delegations` and `UnbondingDelegations` but also each validator can have state bonded, unboned, etc. 

Right now we split `GenesisData.Validators.Delegations` to Bonded `GenesisData.Delegations`, and `GenesisData.UnbondedDelegations` 
And `GenesisData.Validators.UnbondingDelegations` map is used for `GenesisData.UnbondingDelegations`

There can actually be multiple different combinations like:
bonded delegation to bonded validator
unbonding delegation to bonded validator
bonded delegation to unbonding validator
...

I am not sure how to handle this. 

What we currently do is that we withdraw all delegations and we recreate only bonded delegations to bonded validators.